components/capi/src/custom_writeable.rs

```diff
@@ -0,0 +1,97 @@
+use std::ffi::c_void;
+use std::{fmt, ptr};
+
+#[repr(C)]
+/// An object that can one can write UTF-8 strings to
+///
+/// This allows the C API to write to arbitrary kinds of objects, for example a
+/// C++ std::string or a char buffer.
+///
+/// [`icu4x_simple_writeable()`] can be used to write to a fixed-size char buffer.
+///
+/// May be extended in the future to support further invariants
+///
+/// # Safety invariants:
+///  - `flush()` and `grow()` will be passed `data` and the value should be valid for that
+///     `data` may be  null, however `flush()` and `grow()` must then be ready to receive it
+///  - `buf` must be `cap` bytes long
+///  - `grow()` must either return null or a valid buffer of at least the requested buffer size
+///  - Rust code must call `ICU4XCustomWriteable::flush()` before releasing to C
+pub struct ICU4XCustomWriteable {
```

Comment:
Suggestion: Let's bikeshed the name of this struct.  Why did you put "Custom" in the name?

Reply:
Because I copied the name from the doc, I'm not particularly attached to it

Reply:
Why not just `ICU4XWriteable`?

Reply:
works for me!